File: bin/ncbigene_rdf.pl
#!/usr/bin/perl -w
use strict;
use File::Basename;
use Getopt::Std;
my $PROGRAM = basename $0;
my $USAGE=
"Usage: $PROGRAM
";

my %OPT;
getopts('', \%OPT);

print '@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .', "\n";
print '@prefix dct: <http://purl.org/dc/terms/> .', "\n";
print '@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .', "\n";
print '@prefix ncbigene: <http://identifiers.org/ncbigene/> .', "\n";
print '@prefix taxid: <http://identifiers.org/taxonomy/> .', "\n";
print '@prefix hgnc: <http://identifiers.org/hgnc/> .', "\n";
print '@prefix mim: <http://identifiers.org/mim/> .', "\n";
print '@prefix mirbase: <http://identifiers.org/mirbase/> .', "\n";
print '@prefix ensembl: <http://identifiers.org/ensembl/> .', "\n";
print '@prefix insdc: <http://ddbj.nig.ac.jp/ontologies/nucleotide/> .', "\n";
print '@prefix : <https://dbcls.github.io/ncbigene-rdf/ontology.ttl#> .', "\n";

!@ARGV && -t and die $USAGE;

my $ID = "";
while (<>) {
    chomp;
    if (/^#/) {
        print STDERR "Skip $_\n";
        next;
    }

    my @field = split("\t");
    $ID = $field[1];
    my $label = quote_str($field[2]);

    print "\n";
    print "ncbigene:$field[1] a insdc:Gene ;\n";
    print "    dct:identifier \"$field[1]\" ;\n";
    print "    rdfs:label $label ;\n";
    if ($field[10] ne "-") {
        my $standard_name = quote_str($field[10]);
        print "    insdc:standard_name $standard_name ;\n";
    }
    if ($field[3] ne "-") {
        my $locus_tag = quote_str($field[3]);
        print "    insdc:locus_tag $locus_tag ;\n";
    }
    if ($field[4] ne "-") {
        my $synonyms = format_str_array($field[4]);
        print "    insdc:gene_synonym $synonyms ;\n";
    }
    if ($field[8] ne "-") {
        my $description = quote_str($field[8]);
        print "    dct:description $description ;\n";
    }
    if ($field[13] ne "-") {
        my $others = format_str_array($field[13], $field[8]); # exclude redundant description from other_designations
        if ($others) {
            print "    dct:alternative $others ;\n";
        }
    }
    if ($field[5] ne "-") {
        my $link = format_link($field[5]);
        if ($link) {
            print "    insdc:dblink $link ;\n";
        }
    }
    print "    :typeOfGene \"$field[9]\" ;\n";
    if ($field[12] eq "O") {
        print "    :nomenclatureStatus \"official\" ;\n";
    } elsif ($field[12] eq "I") {
        print "    :nomenclatureStatus \"interim\" ;\n";
    }
    if ($field[11] ne "-") {
        print "    :fullName \"$field[11]\" ;\n";
    }
    if ($field[5] ne "-") {
        my $db_xref = filter_str($field[5]);
        if ($db_xref) {
            print "    insdc:db_xref $db_xref ;\n";
        }
    }
    if ($field[15] ne "-") {
        my $feature_type = format_str_array($field[15]);
        print "    :featureType $feature_type ;\n";
    }
    print "    :taxid taxid:$field[0] ;\n";
    if ($field[6] ne "-") {
        print "    insdc:chromosome \"$field[6]\" ;\n";
    }
    if ($field[7] ne "-") {
        print "    insdc:map \"$field[7]\" ;\n";
    }
    my $date = format_date($field[14]);
    print "    dct:modified \"$date\"^^xsd:date .\n";
}

sub format_date {
    my ($date) = @_;
    if ($date =~ /^(\d\d\d\d)(\d\d)(\d\d)$/) {
        my ($y, $m, $d) = ($1, $2, $3);
        return "$y-$m-$d";
    }
}

sub format_str_array {
    my ($str, $exclude) = @_;
    my @arr = split(/\|/, $str);
    my @str_arr = ();
    for my $a (@arr) {
        if ($exclude && $a eq $exclude) {
            next;
        }
        push(@str_arr, quote_str($a));
    }
    if (@str_arr) {
        return join(" ,\n        ", @str_arr);
    }
}

sub format_link {
    my ($str) = @_;
    my @arr = split(/\|/, $str);
    my @link = ();
    for my $a (@arr) {
        if ($a =~ /^MIM:(\d+)$/) {
            push(@link, "mim:$1");
        } elsif ($a =~ /^HGNC:HGNC:(\d+)$/) {
            push(@link, "hgnc:$1");
        } elsif ($a =~ /^Ensembl:(\w+)$/) {
            push(@link, "ensembl:$1");
        } elsif ($a =~ /^miRBase:(MI\d+)$/) {
            push(@link, "mirbase:$1");
        }
    }
    return join(" ,\n        ", @link);
}

sub filter_str {
    my ($str) = @_;
    my @arr = split(/\|/, $str);
    my @link = ();
    for my $a (@arr) {
        if ($a =~ /^MIM:(\d+)$/) {
        } elsif ($a =~ /^HGNC:HGNC:(\d+)$/) {
        } elsif ($a =~ /^Ensembl:(\w+)$/) {
        } elsif ($a =~ /^miRBase:(MI\d+)$/) {
        } else {
            push(@link, quote_str($a));
        }
    }
    return join(" ,\n        ", @link);
}

sub quote_str {
    my ($str) = @_;

    if ($str =~ /\\/) {
        print STDERR "$ID $str";
        $str =~ s/\\/\\\\/g;
        print STDERR " => $str\n";
    }

    my $quoted;
    if ($str =~ /^"[^"]*"$/) {
        $quoted = $str;
        print STDERR "$ID $quoted\n";
    } elsif ($str =~ /"/ && $str =~ /'/ && $str !~ /"""/) {
        $quoted = '"""' . $str . '"""';
        print STDERR "$ID $quoted\n";
    } elsif ($str !~ /"/) {
        $quoted = '"' . $str . '"';
    } elsif ($str !~ /'/) {
        $quoted =  "'$str'";
        print STDERR "$ID $quoted\n";
    } else {
        die $str;
    }
    return $quoted;
}

File: bin/ncbigene_rdf.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead, Write};

fn main() -> std::io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <input_file>", args[0]);
        return Ok(());
    }
    let input_file = &args[1];
    let file = File::open(input_file)?;

    let mut writer = io::BufWriter::new(io::stdout());

    writeln!(writer, "@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .")?;
    writeln!(writer, "@prefix dct: <http://purl.org/dc/terms/> .")?;
    writeln!(writer, "@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .")?;
    writeln!(writer, "@prefix ncbigene: <http://identifiers.org/ncbigene/> .")?;
    writeln!(writer, "@prefix taxid: <http://identifiers.org/taxonomy/> .")?;
    writeln!(writer, "@prefix hgnc: <http://identifiers.org/hgnc/> .")?;
    writeln!(writer, "@prefix mim: <http://identifiers.org/mim/> .")?;
    writeln!(writer, "@prefix mirbase: <http://identifiers.org/mirbase/> .")?;
    writeln!(writer, "@prefix ensembl: <http://identifiers.org/ensembl/> .")?;
    writeln!(writer, "@prefix insdc: <http://ddbj.nig.ac.jp/ontologies/nucleotide/> .")?;
    writeln!(writer, "@prefix : <https://dbcls.github.io/ncbigene-rdf/ontology.ttl#> .")?;

    let reader = io::BufReader::new(file);
    for line in reader.lines() {
        let line = line?;
        if line.starts_with('#') {
            continue;
        }

        let fields: Vec<&str> = line.split('\t').collect();

        writeln!(writer, "")?;
        writeln!(writer, "ncbigene:{} a insdc:Gene ;", fields[1])?;
        writeln!(writer, "    dct:identifier {} ;", fields[1])?;
        writeln!(writer, "    rdfs:label {} ;", quote_str(fields[2]))?;
        if fields[10] != "-" {
            writeln!(writer, "    insdc:standard_name {} ;", quote_str(fields[10]))?;
        }
        if fields[3] != "-" {
            writeln!(writer, "    insdc:locus_tag {} ;", quote_str(fields[3]))?;
        }
        if fields[4] != "-" {
            writeln!(writer, "    insdc:gene_synonym {} ;", format_str_array(fields[4]))?;
        }
        if fields[8] != "-" {
            writeln!(writer, "    dct:description {} ;", quote_str(fields[8]))?;
        }
        if fields[13] != "-" {
            let others = format_str_array_exclude(fields[13], fields[8]);
            if others != "" {
                writeln!(writer, "    dct:alternative {} ;", others)?;
            }
        }
        if fields[5] != "-" {
            let link = format_link(fields[5]);
            if link != "" {
                writeln!(writer, "    insdc:dblink {} ;", link)?;
            }
        }
        writeln!(writer, "    :typeOfGene \"{}\" ;", fields[9])?;
        if fields[12] == "O" {
            writeln!(writer, "    :nomenclatureStatus \"official\" ;")?;
        } else if fields[12] == "I" {
            writeln!(writer, "    :nomenclatureStatus \"interim\" ;")?;
        }
        if fields[11] != "-" {
            writeln!(writer, "    :fullName \"{}\" ;", fields[11])?;
        }
        if fields[5] != "-" {
            let db_xref = filter_str(fields[5]);
            if db_xref != "" {
                writeln!(writer, "    insdc:db_xref {} ;", db_xref)?;
            }
        }
        if fields[15] != "-" {
            writeln!(writer, "    :featureType {} ;", format_str_array(fields[15]))?;
        }
        writeln!(writer, "    :taxid taxid:{} ;", fields[0])?;
        if fields[6] != "-" {
            writeln!(writer, "    insdc:chromosome \"{}\" ;", fields[6])?;
        }
        if fields[7] != "-" {
            writeln!(writer, "    insdc:map \"{}\" ;", fields[7])?;
        }
        writeln!(writer, "    dct:modified \"{}\"^^xsd:date .", format_date(fields[14]))?;
    }
    Ok(())
}

fn quote_str(s: &str) -> String {
    let mut r = String::from(s);
    if r.contains('\\') {
        r = r.replace("\\", "\\\\");
    }

    if r.contains('"') && r.contains('\'') && !r.contains("\"\"\"") {
        format!("\"\"\"{}\"\"\"", r)
    } else if !r.contains('"') {
        format!("\"{}\"", r)
    } else if !r.contains('\'') {
        format!("'{}'", r)
    } else {
        r
    }
}

fn format_str_array(str: &str) -> String {
    let arr: Vec<&str> = str.split('|').collect();
    let str_arr: Vec<String> = arr.iter().map(|a| quote_str(a)).collect();
    str_arr.join(" ,\n        ")
}

fn format_str_array_exclude(str: &str, exclude: &str) -> String {
    let arr: Vec<&str> = str.split('|').collect();
    let mut out: Vec<String> = Vec::new();
    for a in arr {
        if a == exclude {
            continue;
        }
        out.push(quote_str(a));
    }
    out.join(" ,\n        ")
}

fn format_link(str: &str) -> String {
    let arr: Vec<&str> = str.split('|').collect();
    let mut link: Vec<String> = Vec::new();
    for a in arr {
        if let Some(mim) = a.strip_prefix("MIM:") {
            link.push(format!("mim:{}", mim));
        } else if let Some(hgnc) = a.strip_prefix("HGNC:HGNC:") {
            link.push(format!("hgnc:{}", hgnc));
        } else if let Some(ensembl) = a.strip_prefix("Ensembl:") {
            link.push(format!("ensembl:{}", ensembl));
        } else if let Some(mirbase) = a.strip_prefix("miRBase:") {
            link.push(format!("mirbase:{}", mirbase));
        }
    }
    link.join(" ,\n        ")
}

fn filter_str(str: &str) -> String {
    let arr: Vec<&str> = str.split('|').collect();
    let mut link: Vec<String> = Vec::new();
    for a in arr {
        if a.starts_with("MIM:") {
        } else if a.starts_with("HGNC:HGNC:") {
        } else if a.starts_with("Ensembl:") {
        } else if a.starts_with("miRBase:") {
        } else {
            link.push(quote_str(a));
        }
    }
    link.join(" ,\n        ")
}

fn format_date(date: &str) -> String {
    if date.len() == 8 {
        format!("{}-{}-{}", &date[..4], &date[4..6], &date[6..8])
    } else {
        String::new()
    }
}
